src/main/java/org/ibp/api/java/impl/middleware/germplasm/validator/GermplasmNameValidator.java

```diff
@@ -0,0 +1,147 @@
+
+	@Autowired
+	GermplasmValidator germplasmValidator;
+
+	@Autowired
+	GermplasmNameService germplasmNameService;
+
+	public void validate(final GermplasmNameRequestDto germplasmNameRequestDto, final String programUUID) {
+		this.errors = new MapBindingResult(new HashMap<String, String>(), GermplasmNameRequestDto.class.getName());
+		germplasmValidator.validateGermplasmId(errors, germplasmNameRequestDto.getGid());
+
+		if (errors.hasErrors()) {
+			throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+
+		if (germplasmNameRequestDto.getId() != null) {
+			final Name name = germplasmNameService.getNameByNameId(germplasmNameRequestDto.getId());
+			this.ValidateGermplasmName(errors, germplasmNameRequestDto, name);
+
+			if (!StringUtils.isBlank(germplasmNameRequestDto.getNameTypeCode())) {
+				this.validateNameTypeCode(errors, germplasmNameRequestDto);
+			}
+			if (germplasmNameRequestDto.getName() != null) {
+				this.validateNameLength(errors, germplasmNameRequestDto);
+			}
+
+			if (germplasmNameRequestDto.getDate() != null) {
+				this.validateNameDate(errors, germplasmNameRequestDto);
+			}
+
+			if (germplasmNameRequestDto.isPreferredName() != null) {
+				this.validateUpdatePreferredName(errors, germplasmNameRequestDto, name);
+			}
+		} else {
+			this.validateNameTypeCode(errors, germplasmNameRequestDto);
+			this.validateNameLength(errors, germplasmNameRequestDto);
+			this.validateNameDate(errors, germplasmNameRequestDto);
+			this.validatePreferredName(errors,germplasmNameRequestDto);
+		}
+	}
+
+	public void validatePreferredName(final BindingResult errors, final GermplasmNameRequestDto germplasmNameRequestDto) {
+		if(germplasmNameRequestDto.isPreferredName() != null){
+			errors.reject("germplasm.name.preferred.required", "");
+			throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+	}
+
+	public void validateDeleteName(final GermplasmNameRequestDto germplasmNameRequestDto) {
+		this.errors = new MapBindingResult(new HashMap<String, String>(), GermplasmNameRequestDto.class.getName());
+		germplasmValidator.validateGermplasmId(errors, germplasmNameRequestDto.getGid());
+
+		if (errors.hasErrors()) {
+			throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+
+		final Name name = germplasmNameService.getNameByNameId(germplasmNameRequestDto.getId());
+		this.ValidateGermplasmName(errors, germplasmNameRequestDto, name);
+		this.ValidateDeletepreferredName(errors, name);
+
+	}
+
+	protected void ValidateGermplasmName(final BindingResult errors, final GermplasmNameRequestDto germplasmNameRequestDto,
+		final Name name) {
+		if (name == null || !name.getGermplasmId().equals(germplasmNameRequestDto.getGid())) {
+			errors.reject("germplasm.name.invalid", "");
+			throw new ApiRequestValidationException(errors.getAllErrors());
+
+		}
+	}
+
+	protected void ValidateDeletepreferredName(final BindingResult errors, final Name name) {
+		if (name.getNstat().equals(1)) {
```

Comment:
same here, to avoid NPE switch the order

Reply:
done.